flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/PbSchemaValidator.java

```diff
@@ -0,0 +1,148 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.protobuf;
+
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.types.logical.ArrayType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.MapType;
+import org.apache.flink.table.types.logical.RowType;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor.JavaType;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PbSchemaValidator {
+	private Descriptors.Descriptor descriptor;
+	private RowType rowType;
+	private Map<JavaType, List<LogicalTypeRoot>> typeMatchMap = new HashMap();
+
+	public PbSchemaValidator(Descriptors.Descriptor descriptor, RowType rowType) {
+		this.descriptor = descriptor;
+		this.rowType = rowType;
+		typeMatchMap.put(JavaType.BOOLEAN, Collections.singletonList(LogicalTypeRoot.BOOLEAN));
+		typeMatchMap.put(
+			JavaType.BYTE_STRING,
+			Arrays.asList(LogicalTypeRoot.BINARY, LogicalTypeRoot.VARBINARY));
+		typeMatchMap.put(JavaType.DOUBLE, Collections.singletonList(LogicalTypeRoot.DOUBLE));
+		typeMatchMap.put(JavaType.FLOAT, Collections.singletonList(LogicalTypeRoot.FLOAT));
+		typeMatchMap.put(
+			JavaType.ENUM,
+			Arrays.asList(LogicalTypeRoot.VARCHAR, LogicalTypeRoot.CHAR));
+		typeMatchMap.put(
+			JavaType.STRING,
+			Arrays.asList(LogicalTypeRoot.VARCHAR, LogicalTypeRoot.CHAR));
+		typeMatchMap.put(JavaType.INT, Collections.singletonList(LogicalTypeRoot.INTEGER));
+		typeMatchMap.put(JavaType.LONG, Collections.singletonList(LogicalTypeRoot.BIGINT));
+	}
+
+	public Descriptors.Descriptor getDescriptor() {
+		return descriptor;
+	}
+
+	public void setDescriptor(Descriptors.Descriptor descriptor) {
+		this.descriptor = descriptor;
+	}
+
+	public RowType getRowType() {
+		return rowType;
+	}
+
+	public void setRowType(RowType rowType) {
+		this.rowType = rowType;
+	}
+
+	public void validate() {
+		validateTypeMatch(descriptor, rowType);
+		if (!descriptor.getFile().getOptions().getJavaPackage()
+			.equals(descriptor.getFile().getPackage())) {
+			throw new IllegalArgumentException(
+				"java_package and package must be the same in proto definition");
+		}
+		if (!descriptor.getFile().getOptions().getJavaMultipleFiles()) {
```

Comment:
`java_multiple_files` is a java option. Can we support the case with `java_multiple_files = false`?

Reply:
It is the same issue of java_package/package, if java_multiple_files=false, the outer class name will also XXXOuter， also user can set java_outer_class_name to override the outer class name. Mixing with java_package/package, it is complicated to get the real java type of nested message type.

Reply:
If someone can tell me how to get the real java class name from com.google.protobuf.Descriptors.Descriptor, the problem can be solved.

Reply:
The complicated code is below, you can see that "OuterClass" is hard coded.
```
    public static String getJavaFullName(Descriptors.Descriptor descriptor) {
        String javaPackageName = descriptor.getFile().getOptions().getJavaPackage();
        if (descriptor.getFile().getOptions().getJavaMultipleFiles()) {
            //multiple_files=true
            if (null != descriptor.getContainingType()) {
                //nested type
                String parentJavaFullName = getJavaFullName(descriptor.getContainingType());
                return parentJavaFullName + "." + descriptor.getName();
            } else {
                //top level message
                return javaPackageName + "." + descriptor.getName();
            }
        } else {
            //multiple_files=false
            if (null != descriptor.getContainingType()) {
                //nested type
                String parentJavaFullName = getJavaFullName(descriptor.getContainingType());
                return parentJavaFullName + "." + descriptor.getName();
            } else {
                //top level message
                if (!descriptor.getFile().getOptions().hasJavaOuterClassname()) {
                    //user do not define outer class name in proto file
                    return javaPackageName + "." + descriptor.getName() + "OuterClass" + "." + descriptor.getName();
                } else {
                    String outerName = descriptor.getFile().getOptions().getJavaOuterClassname();
                    //user define outer class name in proto file
                    return javaPackageName + "." + outerName + "." + descriptor.getName();
                }
            }
        }
    }
```

Reply:
For now I agree with the current implementation with the same reason of https://github.com/apache/flink/pull/14376#discussion_r555667701

Reply:
Now the PR can support different combinations of java_package, package, java_outer_class_name, java_multiple_files now.